manager/src/main/java/org/openremote/manager/mqtt/DefaultMQTTHandler.java

```diff
@@ -102,65 +77,33 @@ public void start(Container container) throws Exception {
-
-        headers.remove(ConnectionConstants.SESSION_CLOSE);
         headers.put(ConnectionConstants.SESSION_OPEN, true);
 
         // Put a close connection runnable into the headers for the client event service
-        // Get the current session which is specific to the current connection
-        Runnable closeRunnable = mqttBrokerService.getForceDisconnectRunnable(connection.getClientId());
+        Runnable closeRunnable = mqttBrokerService.getForceDisconnectRunnable(connection);
         headers.put(ConnectionConstants.SESSION_TERMINATOR, closeRunnable);
         messageBrokerService.getProducerTemplate().sendBodyAndHeaders(ClientEventService.CLIENT_EVENT_QUEUE, null, headers);
         return true;
     }
 
     @Override
-    public void onDisconnect(MqttConnection connection, InterceptDisconnectMessage msg) {
-        super.onDisconnect(connection, msg);
+    public void onDisconnect(RemotingConnection connection) {
+        super.onDisconnect(connection);
 
         Map<String, Object> headers = prepareHeaders(connection);
         headers.put(ConnectionConstants.SESSION_CLOSE, true);
         messageBrokerService.getProducerTemplate().sendBodyAndHeaders(ClientEventService.CLIENT_EVENT_QUEUE, null, headers);
     }
 
     @Override
-    public void onConnectionLost(MqttConnection connection, InterceptConnectionLostMessage msg) {
-        super.onConnectionLost(connection, msg);
+    public void onConnectionLost(RemotingConnection connection) {
+        super.onConnectionLost(connection);
         Map<String, Object> headers = prepareHeaders(connection);
         headers.put(ConnectionConstants.SESSION_CLOSE_ERROR, true);
         messageBrokerService.getProducerTemplate().sendBodyAndHeaders(ClientEventService.CLIENT_EVENT_QUEUE, null, headers);
-
-        // Try and publish last will if configured
-        if (connection.hasLastWill()) {
-            // Pass through standard canPublish/publish
-            MQTTLastWill lastWill = connection.getLastWill();
-            Topic lastWillTopic = Topic.asTopic(lastWill.getTopic());
-
-            if (handlesTopic(lastWillTopic)) {
-
-                getLogger().info("Processing last will for client: " + connection);
-
-                if (canPublish(connection, lastWillTopic)) {
-                    MqttFixedHeader fixedHeader = new MqttFixedHeader(MqttMessageType.PUBLISH, true, MqttQoS.EXACTLY_ONCE, false, 0);
-                    MqttPublishVariableHeader varHeader = new MqttPublishVariableHeader(lastWill.getTopic(), 0);
-                    ByteBuf payload = Unpooled.wrappedBuffer(lastWill.getPayload().toString().getBytes());
-                    InterceptPublishMessage message = new InterceptPublishMessage(new MqttPublishMessage(fixedHeader, varHeader, payload), msg.getClientID(), msg.getUsername());
-                    doPublish(connection, lastWillTopic, message);
-                }
-            }
-        }
```

Comment:
Last will functionality removed? Wasn't this implemented recently?

Reply:
Got clarification; this is automatically handled by ActiveMQ. Tests are still there to cover it.